wgsl/index.bs

```diff
@@ -1701,57 +1712,74 @@ required alignment of host-shareable type |S| when used by storage class |C|.
   <tr><td>[=i32=], [=u32=], or [=f32=]
       <td>[=AlignOf=](|S|)
       <td>[=AlignOf=](|S|)
-  <tr><td>vec|N|&lt;`T`&gt;
+  <tr><td>atomic&lt;T&gt;
+      <td>[=AlignOf=](|S|)
+      <td>[=AlignOf=](|S|)
+  <tr><td>vecN&lt;T&gt;
       <td>[=AlignOf=](|S|)
       <td>[=AlignOf=](|S|)
   <tr algorithm="alignment of a matrix with N columns and M rows">
-      <td>mat|N|x|M|&lt;f32&gt;
+      <td>matNxM&lt;f32&gt;
       <td>[=AlignOf=](|S|)
       <td>[=AlignOf=](|S|)
   <tr algorithm="alignment of an array">
-      <td>array<|T|,|N|>
-      <td>[=AlignOf=](|T|)
-      <td>[=roundUp=](16, [=AlignOf=](|T|))
+      <td>array&lt;T, N&gt;
+      <td>[=AlignOf=](|S|)
+      <td>[=roundUp=](16, [=AlignOf=](|S|))
   <tr algorithm="alignment of an runtime-sized array">
-      <td>array<|T|>
-      <td>[=AlignOf=](|T|)
-      <td>[=roundUp=](16, [=AlignOf=](|T|))
+      <td>array&lt;T&gt;
+      <td>[=AlignOf=](|S|)
+      <td>[=roundUp=](16, [=AlignOf=](|S|))
   <tr algorithm="alignment of a structure">
-      <td>struct&lt;T<sub>0</sub>, ..., T<sub>N</sub>&gt;
-      <td>max([=AlignOf=](T<sub>0</sub>), ..., [=AlignOf=](T<sub>N</sub>))
-      <td>[=roundUp=](16, max([=AlignOf=](T<sub>0</sub>), ..., [=AlignOf=](T<sub>N</sub>)))<br>
-  <tr><td>atomic<|T|>
-      <td>[=AlignOf=](|T|)
-      <td>[=AlignOf=](|T|)
+      <td>struct |S|
+      <td>[=AlignOf=](|S|)
+      <td>[=roundUp=](16, [=AlignOf=](|S|))<br>
 </table>
 
-All structure members of type |T| must have a byte offset from the start of the
-structure that is a multiple of the [=RequiredAlignOf=](|T|, |C|) for the storage
-class |C|:
+Structure members of type |T| must have a byte offset
+from the start of the structure that is a multiple of the [=RequiredAlignOf=](|T|, |C|)
+for the storage class |C|:
 
 <p algorithm="structure member minimum alignment">
-    [=OffsetOf=](|S|, |M|) = |k| &times; [=RequiredAlignOf=](|T|, C)<br>
+    [=OffsetOfMember=](|S|, |M|) = |k| &times; [=RequiredAlignOf=](|T|, C)<br>
     Where |k| is a non-negative integer and |M| is a member of structure |S| with type |T|
 </p>
 
-All arrays of element type |T| must have an element [=stride=] that is a
-multiple of [=RequiredAlignOf=](|T|, |C|) for the storage class |C|:
+Arrays of element type |T| must have an element [=stride=] that is a
+multiple of the [=RequiredAlignOf=](|T|, |C|) for the storage class |C|:
 
 <p algorithm="array element minimum alignment">
     [=StrideOf=](array<|T|[, |N|]>) = |k| &times; [=RequiredAlignOf=](|T|, C)<br>
     Where |k| is a non-negative integer
 </p>
 
+Note: It is not required for [=AlignOf=](|T|) to be a multiple of [=RequiredAlignOf=](|T|, |C|),
```

Comment:
I'm confused by this note.

The second part seems to say you can provide a too-small align attribute, because the defaults for RequiredAlignmentOf will take over. Also, that the shader author has other kinds of control over the byte offset.
But I thought the point of supplying the attributes is to over-determine the values.

Maybe an example would help?

Reply:
I'm trying to explicitly state that `RequiredAlignOf(T, C)` does not impose any additional restrictions on the values permitted to `[[align]]`, nor does it affect the rules on `AlignOf()`.

Everything is laid out with the rules defined in the previous sections, then for UBO usage, the individual value offsets are checked to make sure that they are correctly aligned to `RequiredAlignOf(T, uniform)`.

So, for example, you can decorate an array with `[[align(4)]]`, even in a UBO, so long as that array ends up being at a 16 byte offset in the buffer.

Clearly this is still not obvious.
I'm not sure how best to describe this.

Reply:
I think your first sentence here is brilliantly on point:

> Note: RequiredAlignOf(T, C) does not impose any additional restrictions on the values permitted for  [[align]], nor does it affect the rules on AlignOf().

Then 

> Data is laid out with the rules defined in previous sections.  Then, when used in a uniform buffer, the resulting layout offsets are must satisfy the RequiredAlignOf(T,uniform) rules.

Reply:
Reworded slightly, but done.